kube-runtime/src/reflector/object_ref.rs

```diff
@@ -1,54 +1,56 @@
 use derivative::Derivative;
 use k8s_openapi::{api::core::v1::ObjectReference, apimachinery::pkg::apis::meta::v1::OwnerReference};
-use kube_client::{
-    api::{DynamicObject, Resource},
-    core::ObjectMeta,
-    ResourceExt,
-};
-use std::{
-    fmt::{Debug, Display},
-    hash::Hash,
-};
+use kube_client::core::{ObjectMeta, Resource, TypeInfo, TypeMeta};
+use std::fmt::{Debug, Display};
+use thiserror::Error;
 
 #[derive(Derivative)]
-#[derivative(
-    Debug(bound = "K::DynamicType: Debug"),
-    PartialEq(bound = "K::DynamicType: PartialEq"),
-    Eq(bound = "K::DynamicType: Eq"),
-    Hash(bound = "K::DynamicType: Hash"),
-    Clone(bound = "K::DynamicType: Clone")
-)]
-/// A typed and namedspaced (if relevant) reference to a Kubernetes object
+#[derivative(Debug, PartialEq, Eq, Hash, Clone)]
+
+/// A dynamically typed reference to an object along with its namespace
 ///
-/// `K` may be either the object type or `DynamicObject`, in which case the
-/// type is stored at runtime. Erased `ObjectRef`s pointing to different types
-/// are still considered different.
+/// Intended to be constructed from one of three sources:
+/// 1. an object returned by the apiserver through the `TypeInfo` trait
+/// 2. an `OwnerReference` found on an object returned by the apiserver
+/// 3. a type implementing `TypeInfo` but with only a `name` pointing to the type
 ///
 /// ```
-/// use kube_runtime::reflector::ObjectRef;
-/// use k8s_openapi::api::core::v1::{ConfigMap, Secret};
-/// assert_ne!(
-///     ObjectRef::<ConfigMap>::new("a").erase(),
-///     ObjectRef::<Secret>::new("a").erase(),
+/// use kube_client::core::Resource;
+/// use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
+/// # use k8s_openapi::api::core::v1::Pod;
+/// # use kube_runtime::reflector::ObjectRef;
+/// let mut pod = Pod::default();
+/// pod.meta_mut().name = Some("foo".into());
+/// let oref = OwnerReference {
+///    api_version: "v1".into(),
+///    kind: "Pod".into(),
+///    name: "foo".into(),
+///    ..OwnerReference::default()
+/// };
+/// assert_eq!(
+///     ObjectRef::from_obj(&pod).within("ns"),
+///     ObjectRef::from_owner(&oref).within("ns"),
+/// );
+/// assert_eq!(
+///     ObjectRef::from_obj(&pod).within("ns"),
+///     ObjectRef::from_resource::<Pod>("foo").within("ns")
```

Comment:
Have reworked the primary ctors of `ObjectRef` such that there are only these three exposed (manual construction still possible) and all of these have a way of injecting the typeinfo by default.